File: solutions/4019-remove-k-balanced-substrings/remove-k-balanced-substrings.py
class Solution:
    def removeSubstring(self, s: str, k: int) -> str:
        open_counts = []
        close_counts = []
        current_open = 0
        current_close = 0
        char_stack = []

        for char in s:
            char_stack.append(char)

            if char == '(':
                if current_close != 0:
                    close_counts.append(current_close)
                    open_counts.append(current_open)
                    current_close = 0
                    current_open = 1
                else:
                    current_open += 1
            elif char == ')':
                current_close += 1

            while True:
                if current_close == k and current_open > k:
                    for _ in range(2 * k):
                        char_stack.pop()
                    current_open -= k
                    current_close -= k
                elif current_close == k and current_open == k:
                    for _ in range(2 * k):
                        char_stack.pop()
                    current_open -= k
                    current_close -= k
                    if char_stack:
                        last_char = char_stack[-1]
                        if last_char == '(':
                            current_open = open_counts.pop()
                        else:
                            current_open = open_counts.pop()
                            current_close = close_counts.pop()
                else:
                    break

        return ''.join(char_stack)

File: solutions/4019-remove-k-balanced-substrings/remove-k-balanced-substrings.java
class Solution {
    public String removeSubstring(String s, int k) {
        List<Integer> openCounts = new ArrayList<>();
        List<Integer> closeCounts = new ArrayList<>();
        int currentOpen = 0;
        int currentClose = 0;
        StringBuilder charStack = new StringBuilder();
        
        for (char c : s.toCharArray()) {
            charStack.append(c);
            
            if (c == '(') {
                if (currentClose != 0) {
                    closeCounts.add(currentClose);
                    openCounts.add(currentOpen);
                    currentClose = 0;
                    currentOpen = 1;
                } else {
                    currentOpen++;
                }
            } else if (c == ')') {
                currentClose++;
            }
            
            while (true) {
                if (currentClose == k && currentOpen > k) {
                    for (int i = 0; i < 2 * k; i++) {
                        charStack.deleteCharAt(charStack.length() - 1);
                    }
                    currentOpen -= k;
                    currentClose -= k;
                } else if (currentClose == k && currentOpen == k) {
                    for (int i = 0; i < 2 * k; i++) {
                        charStack.deleteCharAt(charStack.length() - 1);
                    }
                    currentOpen -= k;
                    currentClose -= k;
                    if (charStack.length() > 0) {
                        char lastChar = charStack.charAt(charStack.length() - 1);
                        if (lastChar == '(') {
                            currentOpen = openCounts.remove(openCounts.size() - 1);
                        } else {
                            currentOpen = openCounts.remove(openCounts.size() - 1);
                            currentClose = closeCounts.remove(closeCounts.size() - 1);
                        }
                    }
                } else {
                    break;
                }
            }
        }
        
        return charStack.toString();
    }
}